### DISCUSSION TITLE
Some query methods are not callable when using the onQuery hook in Lucid relationships

### DISCUSSION THREAD

**Discussion by Obapelumi** [adonisjs/core#1482] [Help] [answered]
Hi,'
Thanks so much for the amazing work on Adonisjs v5. It's been a blessing really.
The Lucid onQuery hook has been super useful and I use it to create several variants of the same relationship. 
However, when I migrated from `@adonisjs/lucid 8.2.2` to `@adonisjs/lucid 8.4.1` some of the query methods on the onQuery hook are no longer callable and some others behave differently from the standard query builder. Some examples are: `apply`, `preload`, `whereIn`  
Here is a sample use case:

```ts
  @manyToMany(() => User, {
    pivotTable: "forum_user",
    onQuery: (query) => query.apply((scopes) => scopes.byRoles(["moderator"])),
  })
  public moderators: ManyToMany<typeof User>;
```
I am working on a forum with users who have different roles. I have a `Forum` model and a `User` model which have a many-to-many relationship between them. There is a `byRoles` scope on the `User` model which filters users based on their roles. In `@adonisjs/lucid 8.2.2` I was able to apply the scope in the onQuery hook but this is no longer allowed.

Would love your insights as regards this change and possible alternatives for my use case.

Thanks again

**Comment by thetutlage** [comment, marked as answer]
Hey @Obapelumi. Yes this happened because now the subqueries used by `withCount`, `whereHas` and `has` methods also invokes this method. You need to wrap your code inside a conditional and it should work fine.

```ts
@manyToMany(() => User, {
  pivotTable: "forum_user",
  onQuery: (query) => {
    if (!query.isRelatedQuery) { return }
    query.apply((scopes) => scopes.byRoles(["moderator"]))
  },
})
public moderators: ManyToMany<typeof User>;
```

I have plans to provide separate hooks for sub-queries and related queries

**Reply by Obapelumi** [reply]
Thanks a lot, @thetutlage. Worked like a charm 💯

**Comment by thetutlage** [comment]
I think this can be fixed. The issue is coming from here https://github.com/Microsoft/TypeScript/issues/571, but can be handled easily. 

Will work on the fix on the weekend

**Reply by Obapelumi** [reply]
Thanks again. I need to dig deeper into the internals of adonisjs and Lucid so maybe I could contribute in the future.
